src/dune_engine/fs_notify_memo.ml

```diff
@@ -8,7 +9,20 @@ let memo =
     ~visibility:Hidden
     (fun _path -> Memo.Build.return ())
 
-let depend path = Memo.exec memo path
+(* Currently, we do not expose this low-level primitive. If you need it, perhaps
+   you could add a higher-level primitive instead, such as [file_exists]? *)
+let depend path =
+  if Path.is_in_build_dir path then
+    Code_error.raise "Fs_notify_memo.depend called on a build path" [];
+  Memo.exec memo path
+
+let file_exists path =
+  let+ () = depend path in
+  Path.exists path
```

Comment:
It probably doesn't matter with how dune currently works (subscribes to everything then filters), but in principle "depend" should happen after the existence check temporally. That way it works even if we make "depend" create an inotify subscription.